### PR TITLE
utils: add method `ensure_formula_installed!`

### PR DESCRIPTION
**Pull request by XuehaiPan**
- [X] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [X] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [X] Have you successfully run `brew style` with your changes locally?
- [X] Have you successfully run `brew typecheck` with your changes locally?
- [X] Have you successfully run `brew tests` with your changes locally?

-----

Reusing common functions, see https://github.com/Homebrew/brew/pull/12465#discussion_r755055468 and https://github.com/Homebrew/brew/pull/12465#discussion_r755159862.

New tests need?

### CODE REVIEW

File: Library/Homebrew/github_packages.rb [hunk 1]
```diff
@@ -53,11 +53,7 @@ def upload_bottles(bottles_hash, keep_old:, dry_run:, warn_on_error:)
       which("skopeo", ENV["HOMEBREW_PATH"]),
       HOMEBREW_PREFIX/"bin/skopeo",
     ].compact.first
```
Comment (MikeMcQuaid) [inline Library/Homebrew/github_packages.rb:55]: I think this logic could be pulled in too to optionally allow using the one from the `PATH` and avoid using `Formula[]` when unneeded?

File: Library/Homebrew/utils.rb [hunk 2]
```diff
@@ -454,6 +454,47 @@ def redirect_stdout(file)
     out.close
   end
 
+  # Ensure the given formula is installed
+  # This is useful for installing a utility formula (e.g. `shellcheck` for `brew style`)
+  def ensure_formula_installed!(formula_or_name, reason = "", latest: false, linked: false,
+                                output_to_stderr: false, quiet: false)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils.rb:460]: I wonder if `output_to_stderr` should just be on by default?

File: Library/Homebrew/utils.rb [hunk 3]
```diff
@@ -454,6 +454,47 @@ def redirect_stdout(file)
...
+
+    require "formula"
+
+    formula = if formula_or_name.is_a?(Formula)
+      formula_or_name
+    else
+      Formula[formula_or_name]
+    end
+
+    reason = " #{reason}" if reason.present? # add a whitespace
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils.rb:481]:
```suggestion
    reason = "for #{reason}" if reason.present? # add a whitespace
```

File: Library/Homebrew/utils.rb [hunk 4]
```diff
@@ -454,6 +454,47 @@ def redirect_stdout(file)
...
+      ohai "Installing `#{formula.name}`#{reason}..."
+      safe_system HOMEBREW_BREW_FILE, "install", "--formula", formula.full_name
+    end
+
+    if latest && !formula.latest_version_installed?
+      ohai "Upgrading `#{formula.name}`#{reason}..."
+      safe_system HOMEBREW_BREW_FILE, "upgrade", "--formula", formula.full_name
+    end
+
+    safe_system HOMEBREW_BREW_FILE, "link", formula.full_name if linked && !formula.linked?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils.rb:493]: I don't think we need to do this, can just use `opt_bin` instead.

**Review by MikeMcQuaid — COMMENTED** [review]
Looks great so far!

File: Library/Homebrew/utils.rb (hunk 3 above)
Comment (XuehaiPan) [reply Library/Homebrew/utils.rb:481]: I add `"for"` in the argument list where this method is called from. Not all `reason` are started with `"for"`.

Comment (MikeMcQuaid) [reply Library/Homebrew/utils.rb:482]: @XuehaiPan I think they could all be reworded to use `for`. Happy to help with suggestions there.

File: Library/Homebrew/style.rb [hunk 5]
```diff
@@ -278,31 +278,14 @@ def shell_scripts
...
-          ohai "Upgrading `shfmt` to format shell scripts..."
-          safe_system HOMEBREW_BREW_FILE, "upgrade", "shfmt"
-        else
-          ohai "Installing `shfmt` to format shell scripts..."
-          safe_system HOMEBREW_BREW_FILE, "install", "shfmt"
-        end
-      end
-
+      ensure_formula_installed!("shfmt", latest: true,
+                                         reason: "to format shell scripts")
```
Comment (MikeMcQuaid) [inline Library/Homebrew/style.rb:288]:
```suggestion
                                         reason: "for formatting shell scripts")
```

File: Library/Homebrew/cmd/update-report.rb [hunk 6]
```diff
@@ -40,12 +40,20 @@ def update_report_args
...
 
   def update_report
+    return update_report_impl if $stdout.tty?
+
+    redirect_stdout($stderr) do
+      return update_report_impl
+    end
+  end
+
+  def update_report_impl
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/update-report.rb:50]:
```suggestion
  def update_report
    return output_update_report if $stdout.tty?

    redirect_stdout($stderr) do
      output_update_report
    end
  end

  private

  def output_update_report
```

File: Library/Homebrew/style.rb [hunk 7]
```diff
@@ -278,31 +278,14 @@ def shell_scripts
 
     def shellcheck
       # Always use the latest brewed shellcheck
```
Comment (MikeMcQuaid) [inline Library/Homebrew/style.rb:280]:
```suggestion
```

File: Library/Homebrew/style.rb [hunk 8]
```diff
@@ -278,31 +278,14 @@ def shell_scripts
...
-        end
-      end
-
-      Formula["shellcheck"].opt_bin/"shellcheck"
+      ensure_formula_installed!("shellcheck", latest: true,
+                                              reason: "shell style checks").opt_bin/"shellcheck"
     end
 
     def shfmt
       # Always use the latest brewed shfmt
```
Comment (MikeMcQuaid) [inline Library/Homebrew/style.rb:286]:
```suggestion
```

File: Library/Homebrew/utils.rb [hunk 9]
```diff
@@ -454,6 +436,60 @@ def redirect_stdout(file)
     out.close
   end
 
+  # Ensure the given formula is installed
+  # This is useful for installing a utility formula (e.g. `shellcheck` for `brew style`)
+  def ensure_formula_installed!(formula_or_name, latest: false, reason: "",
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils.rb:441]:
```suggestion
  def ensure_formula_installed!(formula_or_name, reason: "", latest: false,
```

**Review by MikeMcQuaid — APPROVED** [review]
Some tiny suggestions but: this is great, wonderful work @XuehaiPan!

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
Thanks again @XuehaiPan!

**Pull request merged**